src/realm/object-store/results.cpp

```diff
@@ -1280,6 +1278,18 @@ Results Results::freeze(std::shared_ptr<Realm> const& frozen_realm)
     }
 }
 
+Results Results::freeze(std::shared_ptr<Realm> const& frozen_realm)
+{
+    REALM_ASSERT(frozen_realm->is_frozen());
```

Comment:
This'll force Cocoa to increase the amount of code duplication we have. We currently have just a single resolve(in:) function which is used for both freezing and thawing which calls the OS freeze() functions for both directions.

I don't really see a reason for OS or the c API to have separate free/thaw functions. It's a bunch of extra code that doesn't actually simplify anything on the SDK side.

Reply:
That's a good point. Looking at the OS freeze/thaw implementations, those are just calling `import_copy_of` just changing the name of the arguments. It would make sense to use that across OS and the C API and we can make the freeze/thaw distinction at the SDK layer.

Reply:
We added `freeze/thaw` to avoid having too much implicit logic in the SDK layers. `import_copy_of` is a very internal name that is also easy to misuse if you get the invariants wrong.

Yes, we could just have the logic in the SDK, but the entire point is that OS should be easy to reason about and consume.

It is fair to argue that we shouldn't break existing code, but if Cocoa is using a method called `freeze` to thaw things that sounds like something they should fix regardless.

Reply:
The error handling should also be duplicated and unifying the code would solve that. While I agree `import_copy_of` doesn't roll off the tongue, Cocoa's `resolve` seems good enough.

I don't see much value in having two methods that do the same thing just for the purpose of being explicit. The whole concept of freezing and thawing is a fairly artificial one where all that either method is doing is "find the equivalent object at this version".

Reply:
There isn't really any implicit logic or interesting invariants to maintain. `Realm::freeze()` is the piece that's actually doing the freezing, and I don't see how `obj.freeze(realm.freeze())` and `obj.thaw(realm.thaw())` is any less error prone than `obj.resolve_in(realm.freeze()` and `obj.resolve_in(realm.thaw())`. If anything it seems more because you have to use the correct pair of functions, and I don't see how it gains you anything.

Reply:
Freezing and thawing is not fully symmetric (by intention). Freezing is intended to have a target version in which the object exists. Thawing on the other hand may be to a target version where the object has no counterpart (it has been deleted).

Reply:
OS doesn't enforce any of that, so it's wrong to make that assumption. Using the C API, I can freeze a Realm, create an object in the live Realm, then try to resolve it (freeze it) in the frozen Realm. This is perfectly legal using the C API and will result in an exception, whereas the opposite (thawing a now deleted object) will be handled gracefully.

If anything, the current behaviour of the C API and OS depends on implicit assumptions about how the SDK will consume it.

Reply:
This is an interesting approach @nirinchev . Basically you're saying that you're free to misuse the OS API in any way you want - If it doesn't give you an error, you're good....If nothing else, this re-enforces the need for full error-checking at the OS/SDK boundary.

Reply:
Anyhow - for practical purposes, I think @tgoyne 's proposal of unifying freeze/thaw into a resolve_in() makes sense.

Reply:
Expect that doesn't work with `Realms` themselves. The reason we added `thaw/freeze` everywhere was that it made the API simpler to reason about across all relevant classes as it was the same concept everywhere.

Also to this:

> obj.thaw(realm.thaw())

I would assume that most code is more advanced than that. Kotlin did have bugs due to us passing in the wrong Realm by accident to `import_copy_of`.

IMO one of the goals of OS is to make it easier to consume Core API's. In my mind, that means using the concepts of SDK's, not Core.

Reply:
Freezing and thawing are symmetric, as both freezing and thawing can result in the object not existing.

[The full code for freeze/thaw is pretty much that plus some code to wrap the result in a SDK type](https://github.com/realm/realm-cocoa/blob/a45df7fec01d8f8d1b4ba5ed92a6a9593fde4ac2/Realm/RLMManagedArray.mm#L533). Requiring separate freeze and thaw methods would do nothing but complicate that code with zero upside. Even the C API wrapper in this PR would be simplified by the C++ types not having separate freeze and thaw functions.

Reply:
I agree with @tgoyne for now. Freezing and thawing is currently symmetric. In an ideal world they wouldn't be (as you'd expect a freeze to always succeed - but not have the same expectation for thaw), but we don't know when we can fix that, as it requires substantial changes in Core.